optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java

```diff
@@ -402,20 +403,15 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        // A direct implementation (instead of HashCodeBuilder) to avoid dependencies
-        int hashCode = (17 * 37) + initScore;
-        for (BigDecimal hardScore : hardScores) {
-            hashCode = (37 * hashCode) + hardScore.stripTrailingZeros().hashCode();
-        }
-        for (BigDecimal softScore : softScores) {
-            hashCode = (37 * hashCode) + softScore.stripTrailingZeros().hashCode();
-        }
-        return hashCode;
+        int[] scoreHashCodes = Stream.concat(Arrays.stream(hardScores), Arrays.stream(softScores))
+                .map(BigDecimal::stripTrailingZeros)
+                .mapToInt(BigDecimal::hashCode)
+                .toArray();
+        return Objects.hash(initScore, Arrays.hashCode(scoreHashCodes));
```

Comment:
In my opinion, this is too much to do in `hashCode()`.
In this particular case, I would argue the trailing zeros could be stripped as they come into the `Score`. Precision loss doesn't matter here, as the end value is always going to be the same.

Reply:
The stripTrailingZeros is a requirement of the equals() method, which leaks into the hashcode method due to the Java equal contract.
IIRC, the equals method required that due to this case:
BendableScore.ZERO (= "0") must equal BendableScore.ZERO.add("1.00").subtract("1.00") (= "0.00") or it causes score corruption.

Reply:
That being said, that stripTrailingZeros does look weird, so the equals() method should really comment that example as a motivation so it doesn't get removed by accident.

Reply:
@ge0ffrey My point is... if we strip all trailing zeros always, we don't need to worry about it in `equals()`.